controller/api/destination/watcher/opaque_ports_watcher.go

```diff
@@ -0,0 +1,232 @@
+	OpaquePortsUpdateListener interface {
+		UpdateService(ports map[uint32]struct{})
+	}
+)
+
+// NewOpaquePortsWatcher creates a OpaquePortsWatcher and begins watching for
+// k8sAPI for service changes.
+func NewOpaquePortsWatcher(k8sAPI *k8s.API, log *logging.Entry) *OpaquePortsWatcher {
+	opw := &OpaquePortsWatcher{
+		subscriptions: make(map[ServiceID]*svcSubscriptions),
+		k8sAPI:        k8sAPI,
+		log:           log.WithField("component", "opaque-ports-watcher"),
+	}
+	k8sAPI.Svc().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    opw.addService,
+		DeleteFunc: opw.deleteService,
+		UpdateFunc: func(_, obj interface{}) { opw.addService(obj) },
+	})
+	return opw
+}
+
+// Subscribe subscribes a listener to a service; each time the service
+// changes, the listener will be updated if the list of opaque ports
+// changes.
+func (opw *OpaquePortsWatcher) Subscribe(id ServiceID, listener OpaquePortsUpdateListener) error {
+	opw.Lock()
+	defer opw.Unlock()
+	svc, _ := opw.k8sAPI.Svc().Lister().Services(id.Namespace).Get(id.Name)
+	if svc != nil && svc.Spec.Type == corev1.ServiceTypeExternalName {
+		return invalidService(id.String())
+	}
+	opw.log.Infof("Starting watch on service %s", id)
+	ss, ok := opw.subscriptions[id]
+	// If there is no watched service, create a subscription for the service
+	// and no opaque ports
+	if !ok {
+		opw.subscriptions[id] = &svcSubscriptions{
+			opaquePorts: make(map[uint32]struct{}),
+			listeners:   []OpaquePortsUpdateListener{listener},
+		}
+		return nil
+	}
+	// There are subscriptions for this service, so add the listener to the
+	// service listeners. If there are opaque ports for the service, update
+	// the listener with that value.
+	ss.listeners = append(ss.listeners, listener)
+	if len(ss.opaquePorts) != 0 {
+		listener.UpdateService(ss.opaquePorts)
+	}
+	return nil
+}
+
+// Unsubscribe unsubscries a listener from service.
+func (opw *OpaquePortsWatcher) Unsubscribe(id ServiceID, listener OpaquePortsUpdateListener) {
+	opw.Lock()
+	defer opw.Unlock()
+	opw.log.Infof("Stopping watch on service %s", id)
+	ss, ok := opw.subscriptions[id]
+	if !ok {
+		opw.log.Errorf("Cannot unsubscribe from unknown service %s", id)
+		return
+	}
+	for i, l := range ss.listeners {
+		if l == listener {
+			n := len(ss.listeners)
+			ss.listeners[i] = ss.listeners[n-1]
+			ss.listeners[n-1] = nil
+			ss.listeners = ss.listeners[:n-1]
+		}
+	}
+}
+
+func (opw *OpaquePortsWatcher) addService(obj interface{}) {
+	opw.Lock()
+	defer opw.Unlock()
+	svc := obj.(*corev1.Service)
+	if svc.Namespace == kubeSystem {
+		return
+	}
+	id := ServiceID{
+		Namespace: svc.Namespace,
+		Name:      svc.Name,
+	}
+	opaquePorts, err := getServiceOpaquePortsAnnotation(svc)
+	if err != nil {
+		opw.log.Errorf("failed to get %s service opaque ports annotation: %s", id, err)
+		return
+	}
+	// If the service has no opaque ports there are no updates to send.
+	if len(opaquePorts) == 0 {
```

Comment:
Don't we want to send an update if the opaque ports list changes from non-empty to empty?

Reply:
Yes! We do. Good catch.

I changed the unit tests in `opaque_ports_watcher_test.go` to have an additional `addService` before deletion which acts as an update. I added an additional test that tests this.